consensus/hotstuff/verification/combined_verifier.go

```diff
@@ -84,61 +74,65 @@ func (c *CombinedVerifier) VerifyVote(voterID flow.Identifier, sigData []byte, b
 	}
 
 	// verify each signature against the message
+	// TODO: check if using batch verification is faster (should be yes)
 	stakingValid, err := c.staking.Verify(msg, stakingSig, signer.StakingPubKey)
 	if err != nil {
 		return false, fmt.Errorf("could not verify staking signature: %w", err)
 	}
+	if !stakingValid {
+		return false, nil
+	}
 	beaconValid, err := c.beacon.Verify(msg, beaconShare, beaconPubKey)
 	if err != nil {
 		return false, fmt.Errorf("could not verify beacon signature: %w", err)
 	}
 
-	return stakingValid && beaconValid, nil
+	return beaconValid, nil
 }
 
 // VerifyQC verifies the validity of a combined signature on a quorum certificate.
 func (c *CombinedVerifier) VerifyQC(voterIDs []flow.Identifier, sigData []byte, block *model.Block) (bool, error) {
 
 	// get the full Identities of the signers
 	signers, err := c.committee.Identities(block.BlockID, filter.HasNodeID(voterIDs...))
+	// TODO: only the aggregated public key is needed
+	// TODO: avoid computing the agg public key each time (store the public key delta in CombinedVerifier or c.staking)
 	if err != nil {
 		return false, fmt.Errorf("could not get signer identities: %w", err)
 	}
-	if len(signers) < len(voterIDs) { // check we have valid consensus member Identities for all signers
-		return false, fmt.Errorf("some signers are not valid consensus participants at block %x: %w", block.BlockID, model.ErrInvalidSigner)
+	if len(signers) != len(voterIDs) { // check we have valid consensus member Identities for all signers
+		return false, fmt.Errorf("some signers are not valid consensus participants, or some signers are duplicate, at block %x: %w. signers are %d, voters are %d",
```

Comment:
The convention to wrap errors in our code base is:
* `:` separates the error message of the wrapping error from the error being wrapped
* the wrapped error is the last argument 

This greatly simplifies tracing of the error through the code. 

```suggestion
		return false, fmt.Errorf("duplicate or invalid signer IDs in QC for block %x: %w", block.BlockID, model.ErrInvalidSigner)
```

I don't think we have to give the specific number of invalid / duplicate signers.

Reply:
yep, that's an non intentional mistake